### ISSUE TITLE
Real MITM attack risk in intial key exchanges in countries with extreme Internet surveillance

### ISSUE THREAD

**Issue by mingcenwei** [simplex-chat/simplex-chat#932]
In countries such as China, the government has control over virtually all Internet corporations. Initial key exchanges using links in insecure channels such as email, SMS, or WeChat in these countries are prone to MITM attacks. These platforms may be required to detect SimpleX contact links and to deterministically translate these links into malicious links instantly, and if a malicious link were clicked, the government could then accept the invitation in the corresponding original link and carry out MITM attack.

Of course this problem can be mitigated simply by using secure channels such as Signal or real-time video call, but currently in the documentation of SimpleX the danger of using insecure channels hasn't been emphasized, and many users may simply using insecure channels to send contact links.

**Comment by epoberezkin** [comment]
Interesting. I think we do explain it, but happy to amend the docs to emphasise it.

Some questions:

1) Do you think this is a current problem or the future problem (because of early stage and small user-base size)? That is, would any entity invest into implementing the automatic scalable solution at this stage? This entity would have to implement some custom protocol implementation to proxy communication in real time. It's not impossible, but seems a bit far-fetched at this stage. I'd say it would be a real risk once we cross 1m users, and I don't think such implementation would be even considered before we do – it's non-trivial. I might be wrong though (and we don't account for the users who self-host of course).
2) Beyond improving clarity about this risk, I was actually thinking about adding option for multi-factor link exchange – that is, the link would have to be sent via two (or more) independent channels and all would have to be used in the app (and they would have to match) before the connection is made. What do you think about it? We're not going to prioritise it right now, but when we cross 1m users – might it be worth it?

**Comment by mingcenwei** [comment]
1. I think it’s indeed a future problem, which will be serious when the user base of SimpleX becomes as large as that of Signal now. Nevertheless, I think it only beneficial to add a caveat early on.
2. Multi-factor link exchange will definitely increase security. However, it will be broken if all common platforms use the same deterministic link substitution algorithm, which is possible at the request of some powerful government.

**Comment by epoberezkin** [comment]
> Multi-factor link exchange will definitely increase security. However, it will be broken if all common platforms use the same deterministic link substitution algorithm, which is possible at the request of some powerful government.

Yes, thought about it... We will be adding "connection validation" anyway, for the sake of group connections validation, and the qr/code link exchanged here won't be possible to derive from the initial connection link - that would protect against MITM. So, effectively, the second factor with uncorrelated data.

**Comment by mingcenwei** [comment]
> Yes, thought about it... We will be adding "connection validation" anyway, for the sake of group connections validation, and the qr/code link exchanged here won't be possible to derive from the initial connection link - that should protect against MITM in the scenario you describe. So, effectively, the second factor with uncorrelated data.

If the second factor is different, it will surely increase the cost of automatic MITM attacks hugely!

**Comment by m6n50u7** [comment]
as long as AI doesn't develop enough, there's still possibility to use some crooked ways to exchange the links/keys and such things so that they can't alter them in real time.
